### PR TITLE
console.lua: make completion case-insensitive

### PR DESCRIPTION
**Pull request by guidocella**
This is useful for completing files and more rarely for profiles. It will also be useful to third-party scripts interacting with the console once the API to do it is merged.

### CODE REVIEW

**Comment by Dudemanguy** [comment]
Isn't shell completion usually case sensitive? At least mine is. No strong opinion either way.

**Comment by Dudemanguy** [comment]
Maybe an option? Default it to `auto` which would be case insensitive on windows and case sensitive everywhere else. We already do platform detection in console anyways.

**Comment by guidocella** [comment]
> Isn't shell completion usually case sensitive? At least mine is. No strong opinion either way.

I always used case insensitive completion on Linux, you just have to enable it with `set completion-ignore-case On` in `.inputrc`/`.bashrc` and `setopt  NO_CASE_GLOB` in zsh.

> Maybe an option? Default it to auto which would be case insensitive on windows and case sensitive everywhere else. We already do platform detection in console anyways.

Is there a use case for disabling it? I don't think we should add options if no one needs them.

**Comment by Dudemanguy** [comment]
I mean my filesystem is case sensitive so I would personally prefer my tab completion on those files to be case sensitive too (like it currently is).

**Comment by christoph-heinrich** [comment]
If we had the functionality that people can tab through the suggestions to select one, then I'd say the completion should be case sensitive, but the suggestions should be case insensitive. However as things stand I really can't say what the right behavior should be. If that's how readline behaves, then I guess it can be considered correct :shrug:

File: player/lua/console.lua [hunk 1]
```diff
@@ -811,30 +813,68 @@ function build_completers()
...
+
+    if opts.case_sensitive then
+        return completions, prefix
+    end
+
+    completions = {}
+    local lower_case_completions = {}
+
+    for _, candidate in pairs(list) do
+        if candidate:sub(1, part:len()):lower() == part:lower() then
```
Comment (christoph-heinrich) [inline player/lua/console.lua:856]: Doing `part:lower()` on each iteration ain't great, maybe store it in  a variable instead?

I'm sure it doesn't really matter performance wise in this case, but I'm still bothered by it :P

**Comment by guidocella** [comment]
> If we had the functionality that people can tab through the suggestions to select one, then I'd say the completion should be case sensitive, but the suggestions should be case insensitive. However as things stand I really can't say what the right behavior should be. If that's how readline behaves, then I guess it can be considered correct 🤷

Both bash and zsh can cycle through completions case insensitively, what is the issue?

**Review by christoph-heinrich — APPROVED** [review]
`find_common_prefix()` could be much more efficient by using `common_prefix_length()` instead of creating tons of temporary strings, but that's how that already worked before, so it should really be a separate change.

**Comment by christoph-heinrich** [comment]
> Both bash and zsh can cycle through completions case insensitively, what is the issue?

That we can't do that right now :P

**Comment by guidocella** [comment]
> That we can't do that right now :P

I mean they both allow completing the longest common prefix case insensitevely or cycling case insensitively depending on how you configure them.

**Comment by guidocella** [comment]
> `find_common_prefix()` could be much more efficient by using `common_prefix_length()` instead of creating tons of temporary strings, but that's how that already worked before, so it should really be a separate change.

I did consider reusing `common_prefix_length` but it was simpler to leave it it as it is for now; you can refactor it later if you want.

**Comment by christoph-heinrich** [comment]
> it was simpler to leave it it as it is for now; you can refactor it later if you want.

I wouldn't have given it an "approve" if I thought it should be part of this PR :wink:

File: DOCS/man/console.rst [hunk 2]
```diff
@@ -165,3 +165,8 @@ Configurable Options
 
     The ratio of font height to font width.
     Adjusts table width of completion suggestions.
+
+``case_sensitive``
+    Default: no on Windows, yes on other platforms.
+
+    Whether Tab completion is case sensitive.
```
[the remark this answers is not in the document]
Comment (guidocella) [reply DOCS/man/console.rst]: Nice catch, I added a note.

**Review by Dudemanguy — APPROVED** [review]
LGTM

**Pull request merged**